File: constructor/directory.py
#
# Copyright (c) 2013 Kimball Thurston
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
# OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

import os
import errno
from .dependency import Dependency
from .output import Info, Debug, Error
from .utility import iterate

_curdir = None
def CurDir():
    global _curdir
    return _curdir

def SetCurDir( d ):
    global _curdir
    _curdir = d

def CurBinaryPath():
    global _curdir
    if _curdir:
        return _curdir.bin_path
    Error( "Binary Path not yet configured" )

def CurFullSourcePath():
    return CurDir().src_path

def CurRelativeSourcePath():
    return CurDir().rel_src_path

class Directory(Dependency):
    def __init__( self, path, relpath, binpath, pardir = None, globs = None ):
        super(Directory, self).__init__()
        self.pardir = pardir
        self.globs = globs
        self._cur_namespace = None
        self.src_path = path
        self.rel_src_path = relpath
        self.bin_path = binpath
        # synonym function to handle plurality
        self.add_target = self.add_targets
        # Stuff that the user really turns on / triggers via directives
        self.subdirs = {}
        self.targets = None
        self.modules = None
        self.rules = None
        self.variables = None

    def enable_module( self, mod ):
        if self.modules is None:
            self.modules = {}
        self.modules[mod.name] = mod
        if self._cur_namespace is not None:
            mod.add_globals( self._cur_namespace, "config" )
        mod.set_variables( self )

    def add_module_features( self, driver ):
        if self.modules is not None:
            for n, m in iterate( self.modules ):
                if m.features is not None:
                    for f in m.features:
                        driver.add_feature( f["name"], f["type"], f["help"], f["default"] )
        for sn, sd in iterate( self.subdirs ):
            sd.add_module_features( driver )

    def get_used_module_rules( self, recurse ):
        rl = []
        if self.modules is not None:
            for n, m in iterate( self.modules ):
                if m.rules is not None:
                    for tag, rule in iterate( m.rules ):
                        if rule.is_used():
                            rl.append( rule )
        if recurse:
            for sn, sd in iterate( self.subdirs ):
                subrl = sd.get_used_module_rules( True )
                rl.extend( subrl )
        return rl

    def get_variable( self, varname, expand=True ):
        if self.variables:
            lu = self.variables.get( varname )
            if expand:
                recurval = '$' + varname
                if lu and self.pardir and recurval in lu:
                    parlu = self.pardir.get_variable( varname, expand )
                    if parlu:
                        for x in range( 0, len(lu) ):
                            if lu[x] == recurval:
                                lu[x:x+1] = parlu
        if lu:
            return lu
        if self.pardir:
            return self.pardir.get_variable( varname, expand )
        return None

    def set_variable( self, varname, f ):
        if self.variables is None:
            self.variables = { varname: f }
        else:
            self.variables[varname] = f

    def add_to_variable( self, varname, f ):
        if not isinstance( f, list ):
            if isinstance( f, tuple ):
                tl = []
                for x in f:
                    tl.append( x )
                f = tl
            else:
                f = [ f ]

        if self.variables is None:
            self.variables = { varname: f }
        else:
            v = self.variables.get( varname )
            if v is None:
                self.variables[varname] = f
            else:
                for x in f:
                    v.append( x )

    def transform_variable( self, name, val ):
        if self.modules is not None:
            for n, m in iterate( self.modules ):
                val = m.transform_variable( name, val )
        if self.pardir is not None:
            val = self.pardir.transform_variable( name, val )
        return val

    def set_globals( self, globs ):
        self.globs = globs

    def get_globals( self, phase=None ):
        retval = self.globs
        if retval is None:
            retval = self.pardir.get_globals( phase )

        if phase is None:
            return retval

        retval = retval.copy()
        if self.modules is not None:
            for k, v in iterate( self.modules ):
                v.add_globals( retval, phase )
        self._cur_namespace = retval
        return retval

    def set_bin_dir( self, path ):
        self.bin_path = path
        if len(self.rel_src_path) > 0:
            self.bin_path = os.path.join( path, self.rel_src_path )
        for sn, sd in iterate( self.subdirs ):
            sd.set_bin_dir( path )

    def get_root_bin_dir( self ):
        if self.pardir is None:
            return self.bin_path
        return self.pardir.get_root_bin_dir()

    def add_targets( self, *targs ):
        if self.targets is None:
            self.targets = []

        for t in targs:
            if isinstance( t, list ):
                for x in t:
                    self.targets[:] = [y for y in self.targets if y is not x]
                    self.targets.append( x )
            else:
                self.targets[:] = [x for x in self.targets if x is not t]
                self.targets.append( t )

    def add_sub_dir( self, name ):
        try:
            dobj = self.subdirs[name]
        except KeyError as e:
            newd = self.rel_src_path
            if len(newd) > 0:
                newd = os.path.join( newd, name )
            else:
                newd = name

            dobj = Directory( os.path.join( self.src_path, name ),
                              newd,
                              os.path.join( self.bin_path, name ),
                              self )

            if self.bin_path:
                dobj.set_bin_dir( self.get_root_bin_dir() )

            self.add_dependency( dobj )
            self.subdirs[name] = dobj

        return dobj

    def add_to_globals( self, phase, globnames, locnames ):
        if self.globs is None:
            self.globs = self.get_globals().copy()

        phaseglobs = self.get_globals( phase )
        new_globs = {}
        def _search_globs( ns ):
            all_names = ns.get( "__all__" )
            if all_names is None:
                for k, v in iterate( ns ):
                    if k[0] != '_' and k not in phaseglobs:
                        Debug( "Adding '%s' to globals" % k )
                        new_globs[k] = v
            else:
                for k in all_names:
                    if k not in phaseglobs:
                        Debug( "Adding '%s' to globals" % k )
                        new_globs[k] = ns[k]
        _search_globs( globnames )
        _search_globs( locnames )
        self.globs.update( new_globs )

File: src/Directory.cpp
//
// Copyright (c) 2014 Kimball Thurston
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
// OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

#include "Directory.h"

#include "ScopeGuard.h"
#include "StrUtil.h"
#include "Debug.h"

#include <unistd.h>
#include <errno.h>
#include <system_error>
#include <sys/stat.h>
#include <sys/param.h>
#include <stdlib.h>
#include <sstream>
#include <iostream>
#include <fstream>
#include <list>
#include <iterator>
#include <stack>
#include <mutex>

namespace 
{

static std::once_flag theNeedCWDInit;

static std::string theCWD;
static std::vector<std::string> theCWDPath;

static void initCWD( void )
{
	char *cwd = getcwd( NULL, MAXPATHLEN );
	if ( ! cwd )
		throw std::system_error( errno, std::system_category(),
								 "Unable to query current directory" );
	ON_EXIT{ free( cwd ); };

	theCWD = cwd;
	theCWDPath = String::split( theCWD, File::pathSeparator() );
}

std::shared_ptr<Directory> theLastDir;
static std::stack< std::shared_ptr<Directory> > theLiveDirs;

} // empty namespace


////////////////////////////////////////


Directory::Directory( void )
{
	std::call_once( theNeedCWDInit, &initCWD );

	myFullDirs = theCWDPath;
	myCurFullPath = theCWD;
}


////////////////////////////////////////


Directory::Directory( const std::string &root )
{
	myFullDirs = String::split( root, File::pathSeparator() );
	myCurFullPath = root;
}


////////////////////////////////////////


Directory::Directory( const Directory &d )
		: mySubDirs( d.mySubDirs ),
		  myFullDirs( d.myFullDirs ),
		  myCurFullPath( d.myCurFullPath )
{
}


////////////////////////////////////////


Directory::Directory( Directory &&d )
		: mySubDirs( std::move( d.mySubDirs ) ),
		  myFullDirs( std::move( d.myFullDirs ) ),
		  myCurFullPath( std::move( d.myCurFullPath ) )
{
}


////////////////////////////////////////


Directory &
Directory::operator=( const Directory &d )
{
	if ( this != &d )
	{
		mySubDirs = d.mySubDirs;
		myFullDirs = d.myFullDirs;
		myCurFullPath = d.myCurFullPath;
	}
	return *this;
}


////////////////////////////////////////


Directory &
Directory::operator=( Directory &&d )
{
	if ( this != &d )
	{
		mySubDirs = std::move( d.mySubDirs );
		myFullDirs = std::move( d.myFullDirs );
		myCurFullPath = std::move( d.myCurFullPath );
	}
	return *this;
}


////////////////////////////////////////


void
Directory::extractDirFromFile( const std::string &fn )
{
	myFullDirs = String::split( fn, File::pathSeparator() );
	if ( ! myFullDirs.empty() )
		myFullDirs.pop_back();
	mySubDirs.clear();
	updateFullPath();
}


////////////////////////////////////////


Directory
Directory::reroot( const std::string &newroot ) const
{
#ifdef WIN32
	throw std::runtime_error( "Need to handle UNC paths and / and \\ in split" );
#endif
	Directory ret( newroot );
	ret.mySubDirs = mySubDirs;
	ret.updateFullPath();
	return ret;
}


////////////////////////////////////////


std::shared_ptr<Directory>
Directory::reroot( const std::shared_ptr<Directory> &newroot ) const
{
	std::shared_ptr<Directory> ret = std::make_shared<Directory>( *newroot );
	ret->mySubDirs = mySubDirs;
	ret->updateFullPath();
	return ret;
}


////////////////////////////////////////


void
Directory::rematch( const Directory &d )
{
	bool needChange = true;
	if ( d.mySubDirs.size() == mySubDirs.size() )
	{
		needChange = false;
		for ( size_t i = 0; i != mySubDirs.size(); ++i )
		{
			if ( mySubDirs[i] != d.mySubDirs[i] )
			{
				needChange = true;
				break;
			}
		}
	}
	if ( needChange )
	{
		mySubDirs = d.mySubDirs;
		updateFullPath();
	}
}


////////////////////////////////////////


void
Directory::cd( const std::string &name )
{
	std::vector<std::string> dirs = String::split( name, File::pathSeparator() );
	if ( dirs.empty() )
		return;

	for ( auto &d: dirs )
		mySubDirs.emplace_back( std::move( d ) );

	updateFullPath();
}


////////////////////////////////////////


void
Directory::cdUp( void )
{
	if ( ! mySubDirs.empty() )
		mySubDirs.pop_back();
	else
	{
		if ( myFullDirs.empty() )
			throw std::runtime_error( "Attempt to change directories above root" );
		myFullDirs.pop_back();
	}
	
	updateFullPath();
}


////////////////////////////////////////


const std::string &
Directory::cur( void ) const
{
	if ( ! mySubDirs.empty() )
		return mySubDirs.back();

	if ( ! myFullDirs.empty() )
		return myFullDirs.back();

	return String::empty();
}


////////////////////////////////////////


void
Directory::mkpath( void ) const
{
#ifdef WIN32
	throw std::logic_error( "NYI: Need to handle drive letters" );
#endif
	std::string tmpPath;
	std::vector<std::string> alldirs;
	combinePath( alldirs );
	for ( const auto &curDir: alldirs )
	{
		tmpPath.push_back( File::pathSeparator() );
		tmpPath.append( curDir );

		if ( mkdir( tmpPath.c_str(), 0777 ) != 0 )
		{
			if ( errno == EEXIST )
				continue;

			throw std::system_error( errno, std::system_category(),
				"Unable to create directory '" + tmpPath + "'" );
		}
	}
}


////////////////////////////////////////


const std::string &
Directory::fullpath( void ) const
{
	return myCurFullPath;
}


////////////////////////////////////////


std::string
Directory::relpath( void ) const
{
	std::string ret;

	for ( const std::string &p: mySubDirs )
	{
		if ( ! ret.empty() )
			ret.push_back( File::pathSeparator() );
		ret.append( p );
	}
	return ret;
}


////////////////////////////////////////


void
Directory::promoteFull( void )
{
	for ( auto &d: mySubDirs )
		myFullDirs.emplace_back( std::move( d ) );
	mySubDirs.clear();
}


////////////////////////////////////////


void
Directory::updateFullPath( void )
{
	std::vector<std::string> pathElements;
	combinePath( pathElements );

#ifdef WIN32
	throw std::logic_error( "NYI: Need to handle drive letters" );
#endif
	myCurFullPath.clear();
	for ( const std::string &i: pathElements )
	{
		myCurFullPath.push_back( File::pathSeparator() );
		myCurFullPath.append( i );
	}
}


////////////////////////////////////////


bool
Directory::find( std::string &concatpath, const std::vector<std::string> &names ) const
{
	for ( const auto &n: names )
	{
		if ( exists( concatpath, n ) )
			return true;
	}

	return false;
}


////////////////////////////////////////


bool
Directory::exists( std::string &concatpath, const char *fn ) const
{
	concatpath = makefilename( fn );

	struct stat sb;
	if ( stat( concatpath.c_str(), &sb ) == 0 )
		return true;

	return false;
}


////////////////////////////////////////


bool
Directory::exists( std::string &concatpath, const std::string &fn ) const
{
	concatpath = makefilename( fn );

	struct stat sb;
	if ( stat( concatpath.c_str(), &sb ) == 0 )
		return true;

	return false;
}


////////////////////////////////////////


std::string
Directory::makefilename( const char *fn ) const
{
	std::string concatpath = fullpath();
	concatpath.push_back( File::pathSeparator() );
	concatpath.append( fn );
	return concatpath;
}


////////////////////////////////////////


std::string
Directory::makefilename( const std::string &fn ) const
{
	std::string concatpath = fullpath();
	concatpath.push_back( File::pathSeparator() );
	concatpath.append( fn );
	return concatpath;
}


////////////////////////////////////////


std::string
Directory::relfilename( const std::string &fn ) const
{
	std::string cpath = relpath();
	cpath.push_back( File::pathSeparator() );
	cpath.append( fn );
	return cpath;
}


////////////////////////////////////////


std::string
Directory::relativeTo( const Directory &o,
					   const std::string &fn ) const
{
	std::vector<const std::string *> mynonCommonSubdirs;
	mynonCommonSubdirs.reserve( myFullDirs.size() + mySubDirs.size() );
	for ( const std::string &i: myFullDirs )
		mynonCommonSubdirs.push_back( &i );
	for ( const std::string &i: mySubDirs )
		mynonCommonSubdirs.push_back( &i );

	std::vector<const std::string *> relPath;
	relPath.reserve( o.myFullDirs.size() + o.mySubDirs.size() );
	for ( const std::string &i: o.myFullDirs )
		relPath.push_back( &i );
	for ( const std::string &i: o.mySubDirs )
		relPath.push_back( &i );

	size_t subdirI = 0;
	size_t relI = 0;
	while ( true )
	{
		if ( subdirI == mynonCommonSubdirs.size() || relI == relPath.size() )
			break;
		if ( *(mynonCommonSubdirs[subdirI]) != *(relPath[relI]) )
			break;

		++subdirI;
		++relI;
	}

	std::string ret;
	bool notfirst = false;
	while ( relI < relPath.size() )
	{
		if ( notfirst )
			ret.push_back( File::pathSeparator() );
		ret.append( ".." );
		++relI;
		notfirst = true;
	}
	notfirst = false;
	while ( subdirI < mynonCommonSubdirs.size() )
	{
		if ( notfirst )
			ret.push_back( File::pathSeparator() );
		ret.append( *(mynonCommonSubdirs[subdirI]) );
		++subdirI;
		notfirst = true;
	}

	if ( ! fn.empty() )
	{
		if ( ! ret.empty() )
			ret.push_back( File::pathSeparator() );
		ret.append( fn );
	}

	return ret;
}


////////////////////////////////////////


void
Directory::updateIfDifferent( const std::string &name, const std::vector<std::string> &lines )
{
	bool doCreate = true;
	std::string fn;
	if ( exists( fn, name ) )
	{
		std::ifstream inf( fn );
		size_t curIdx = 0;
		std::string curLine;
		do
		{
			if ( std::getline( inf, curLine ) )
			{
				if ( curIdx >= lines.size() )
					break;

				if ( lines[curIdx] != curLine )
				{
					VERBOSE( name << ": line " << (curIdx + 1) << " differs: '" << curLine << "' vs '" << lines[curIdx] << "' - regenerating" );
					break;
				}
				++curIdx;
			}
			else if ( curIdx == lines.size() )
			{
				doCreate = false;
				break;
			}
			else
			{
				VERBOSE( name << ": line count different - regenerating" );
				break;
			}
		} while ( doCreate );
	}
	
	if ( doCreate )
	{
		mkpath();
		fn = makefilename( name );
		VERBOSE( "Creating/updating '" << fn << "'..." );
		std::ofstream outf( fn );
		for ( const std::string &l: lines )
			outf << l << '\n';
	}
}


////////////////////////////////////////


const std::shared_ptr<Directory> &
Directory::current( void )
{
	if ( theLiveDirs.empty() )
		theLiveDirs.push( std::make_shared<Directory>() );

	return theLiveDirs.top();
}


////////////////////////////////////////


const std::shared_ptr<Directory> &
Directory::pushd( const std::string &d )
{
	std::shared_ptr<Directory> newD = std::make_shared<Directory>( *current() );
	newD->cd( d );
	theLastDir = current();
	theLiveDirs.push( newD );
	return theLiveDirs.top();
}


////////////////////////////////////////


const std::shared_ptr<Directory> &
Directory::popd( void )
{
	if ( theLiveDirs.empty() )
		throw std::runtime_error( "Directory pushd / popd mismatch" );

	theLastDir = theLiveDirs.top();
	theLiveDirs.pop();

	if ( theLiveDirs.empty() )
		throw std::runtime_error( "Directory pushd / popd mismatch" );

	return theLiveDirs.top();
}


////////////////////////////////////////


const std::shared_ptr<Directory> &
Directory::last( void )
{
	return theLastDir;
}


////////////////////////////////////////


void
Directory::combinePath( std::vector<std::string> &elements ) const
{
	elements = myFullDirs;
	for ( const std::string &p: mySubDirs )
	{
		if ( p == "." )
			continue;

		if ( p == ".." )
		{
			if ( elements.empty() )
				throw std::runtime_error( "Invalid attempt to create relative path above root" );

			elements.pop_back();
		}
		else
			elements.push_back( p );
	}

}


////////////////////////////////////////



